File: 2_year_2_semester/spo/football-championship-system/backend/app/schemas/team.py
from typing import List, Optional
from pydantic import BaseModel, validator
from datetime import datetime

class TeamBase(BaseModel):
    name: str
    city: str
    coach: str
    last_season_place: int
    
    @validator('last_season_place')
    def validate_place(cls, v):
        if v < 1 or v > 20:
            raise ValueError('Place must be between 1 and 20')
        return v

class TeamCreate(TeamBase):
    pass

class TeamUpdate(BaseModel):
    name: Optional[str] = None
    city: Optional[str] = None
    coach: Optional[str] = None
    last_season_place: Optional[int] = None
    
    @validator('last_season_place')
    def validate_place(cls, v):
        if v is not None and (v < 1 or v > 20):
            raise ValueError('Place must be between 1 and 20')
        return v

class TeamInDBBase(TeamBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True

class Team(TeamInDBBase):
    pass

class TeamWithPlayers(TeamInDBBase):
    players: List['Player'] = []
    
    class Config:
        orm_mode = True

# Avoid circular imports
from app.schemas.player import Player
TeamWithPlayers.update_forward_refs()

File: 2_year_2_semester/spo/football-championship-system/frontend/src/types/team.ts
// frontend/src/types/team.ts

/**
 * Основной интерфейс для команды, получаемый от бэкенда.
 * Соответствует схеме `Team` в Pydantic.
 */
export interface Team {
  id: number;
  name: string;
  city: string;
  coach: string;
  last_season_place: number;
  created_at: string; // Даты из API приходят как строки в формате ISO
  updated_at: string;
}

/**
 * Интерфейс для создания новой команды.
 * Соответствует схеме `TeamCreate` в Pydantic.
 */
export interface TeamCreate {
  name: string;
  city: string;
  coach: string;
  last_season_place: number;
}

/**
 * Интерфейс для обновления существующей команды.
 * Поля необязательные, соответствует схеме `TeamUpdate` в Pydantic.
 */
export interface TeamUpdate {
  name?: string;
  city?: string;
  coach?: string;
  last_season_place?: number;
}